### PR TITLE
TypeInfo_Struct: Switch to stored mangled name & demangle lazily (with per-thread cache)

### PR DESCRIPTION
**Pull request by kinke**
Making the `TypeInfo_Struct` names truly unique and more compact at the same time. Requires compiler support: https://github.com/dlang/dmd/pull/12928

Caching the demangled name in the instance itself isn't possible because `TypeInfo.toString()` is `const`. I had to remove its `pure` in order to cache the demangled name in a per-thread AA (and using `const void*` as key, as `const TypeInfo_Struct` leads to an infinite recursion...).

### CODE REVIEW

**Comment by kinke** [comment]
I've run the tests locally on Linux x64 with both dmd and druntime branches; all green (dmd/druntime/phobos) except for (apparently hanging) druntime integration subtest `printf` of `test/gc/`.

File: src/rt/util/typeinfo.d
Comment (kinke) [inline src/rt/util/typeinfo.d:471]: `TypeInfo.toString()` isn't pure anymore.

**Comment by rainers** [comment]
> except for (apparently hanging) druntime integration subtest `printf` of `test/gc/`

The test is trying to print the types of allocations while the GC lock is held: this deadlocks when accessing `name`, see https://github.com/dlang/druntime/blob/master/src/core/internal/gc/impl/conservative/gc.d#L4061.

That also shows another subtle change: using `TypeInfo_Struct.name` no longer is `@nogc`. Could you use `HashTab` from `core.internal.container.hashtab` instead of the AA?

**Comment by PetarKirov** [comment]
> TypeInfo.toString() isn't pure

This is a breaking change (although I was expecting quite more to be broken than just `atilaneves/unit-threaded`), it needs a changelog entry and stronger justification.

>  Making the TypeInfo_Struct names truly unique

While this certainly is a good thing, can I explain why is it necessary? Are you trying to solve some bigger problem with this?

**Comment by kinke** [comment]
> The test is trying to print the types of allocations while the GC lock is held: this deadlocks when accessing name

Cheers; worked around for it now by using `mangledName` there.

> That also shows another subtle change: using TypeInfo_Struct.name no longer is @nogc. Could you use HashTab from core.internal.container.hashtab instead of the AA?

The AA shouldn't really matter, as `core.demangle` GC-allocates anyway, unless using some huge buffer that is guaranteed to suffice for all names and then malloc'ing an appropriate copy afterwards...

**Comment by kinke** [comment]
@PetarKirov: I haven't found an existing issue yet, so filed a new one: https://issues.dlang.org/show_bug.cgi?id=22149

This is especially bad for classes (tackled later), where it can e.g. break EH with LDC on Windows (see https://github.com/ldc-developers/ldc/pull/3517), and dynamic casting in general AFAIK.

**Comment by kinke** [comment]
Testcase added to dlang/dmd#12928.

**Comment by PetarKirov** [comment]
Indeed, that's a quite severe. Thanks for the explanations and linking the LDC PR!

**Comment by rainers** [comment]
> Cheers; worked around it for now by using `mangledName` there.

Fine with me, a log can be run through ddemangle if needed.

> The AA shouldn't really matter, as `core.demangle` GC-allocates anyway

Noticed that too after commenting. An alternative could be to abort demangling if the passed buffer is too small and let the caller allocate a larger one. If only the demangler would not use exceptions (which it only needs for backward compatibilty with older ambiguous manglings IIRC). Changing the demangler would be another PR, though.

**Comment by kinke** [comment]
> Fine with me, a log can be run through ddemangle if needed.

`ddemangle` unfortunately doesn't seem to support such type names (`S3foo3bar`).

> An alternative could be to abort demangling if the passed buffer is too small and let the caller allocate a larger one.

Yeah, or passing some sink which is responsible for increasing the internal capacity as needed.

> If only the demangler would not use exceptions (which it only needs for backward compatibilty with older ambiguous manglings IIRC).

Yeah, I find it also very suboptimal as demangling is used for exception backtraces, and throws at that time aren't really to be expected.

> Changing the demangler would be another PR, though.

Definitely. I've just seen that there are other 'beautiful' related hacks like https://github.com/dlang/druntime/blob/3a9ede069874379ee194b8b479941e329d03b68f/src/object.d#L1209-L1210

---

CI doesn't seem to really work with feature branches (same-named branches directly on the official dlang repos) to tackle compiler and druntime at the same time; both PRs currently include a temporary commit to make Cirrus CI use the correct branches at least. Those commits obviously need to be removed before merging.

Restoring `pure` to `TypeInfo_Class.toString()` etc. made unit-threaded pass again for Buildkite. A changelog entry still needs to be added to emphasize that breaking change; not sure if here or in the DMD PR. The latter also includes the (trivial) fix for `TypeInfo_Class` now.

File: src/object.d
Comment (kinke) [inline src/object.d:1712]: `TypeInfo.opCmp` uses `toString()` too; I'm not convinced overriding it here to check for a `TypeInfo_Struct` rhs and comparing the `mangledName` instead is worth it.

**Comment by kinke** [comment]
> CI doesn't seem to really work with feature branches (same-named branches directly on the official dlang repos) to tackle compiler and druntime at the same time

I've made it work for Azure Pipelines now (one according commit in each PR); the druntime patch here is bit of an ugly copy-pasta but does the job. So PRs originating from branches on the official dlang repos will try to use same-named branches for the other repos, and fall back to `master` if the branch doesn't exist. To make that work for Phobos too, it would need a commit analogous to the one here, but Phobos isn't as tightly tied to the compiler as druntime, so I don't think it's really needed. So
* useful when a change involves both compiler and druntime at the same time like this one here,
* only for people with write access to both dlang repos,
* preventing PRs originating from the official repo from targeting `stable` (due to the `master` fallback) - submit those from your fork instead.

**Comment by PetarKirov** [comment]
> CI doesn't seem to really work with feature branches

We really should stop wasting everyone's time with the artificial split between dmd and druntime (and possibly phobos) and merge them in a mono repo.

**Comment by kinke** [comment]
[Would be nice to have this land in 2.098; Walter has approved of the DMD PR.]

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
@kinke It seems that the autotester is failing a couple of tests in dmd. Is this expected?

**Comment by kinke** [comment]
@RazvanN7: Yes, these 2 DMD testsuite failures are expected if DMD and druntime aren't in sync.

**Comment by RazvanN7** [comment]
I'm gonna pull the trigger on these PRs. Let's hope everything will run smoothly.

**Comment by RazvanN7** [comment]
Hmm, unfortunately I don't have the right to merge this unless it passes all required tests. However, I have enough permissions to merge the dmd PR.

Maybe I can do that and wait for this PR to be green? That will stall the entire pipeline. But I don't see any other way out of it.

**Comment by kinke** [comment]
Can't you merge manually via git cmdline? Should I try to?

**Comment by RazvanN7** [comment]
Feel free to merge both PRs. (I apparently can't merge this one)

**Comment by kinke** [comment]
@RazvanN7: I can't merge any one of these PRs either; looks like the GitHub branch protection rules even apply to manual pushes. So if you can indeed merge the DMD one, please go ahead.

**Comment by RazvanN7** [comment]
@kinke Merged it. Maybe re-push to restart failing tests.

**Pull request merged**
